descqa/NumberDensityVersusRedshift.py

```diff
@@ -0,0 +1,307 @@
+data_info = {
+    'Coil2004_magbin': {'filename_template': 'N_z/DEEP2/Coil_et_al_2004_Table3_{}.txt', 'usecols':[0,1,2,4], 'skiprows':2},
+    'Coil2004_maglim': {'filename_template': 'N_z/DEEP2/Coil_et_al_2004_Table4_{}.txt', 'usecols':[0,1,2], 'skiprows':3},
+    'DEEP2_JAN':{'filename_template': 'N_z/DEEP2/JANewman_{}.txt','usecols':[0,1,2,3], 'skiprows':1},
+}
+
+#plotting constants
+figx_p = 9
+figy_p = 11
+lw2 = 2
+fsize = 16
+lsize = 10  
+default_colors = ['blue','r','m','g','navy','y','purple','gray','c','orange','violet','coral','gold','orchid','maroon','tomato','sienna','chartreuse','firebrick','SteelBlue']
+validation_color = 'black'
+
+class NumberDensityVersusRedshift(BaseValidationTest):
+    """
+    validation test to show N(z) distributions
+    """
+    def __init__(self, z='redshift_true', band='i', N_zbins=44,zlo=0., zhi=1.1, observation='', m_lo=27, m_hi=18, ncolumns=2, normed=True, **kwargs):
+        
+        #catalog quantities
+        self.z = z
+        possible_mag_fields = ('mag_{}_lsst',
+                               'mag_{}_sdss',
+                               'mag_{}_des',
+                              )
+        self.possible_mag_fields = [f.format(band) for f in possible_mag_fields]
+        self.band = band
+        
+        #z-bounds and binning
+        self.zlo = zlo
+        self.zhi = zhi
+        self.N_zbins = N_zbins
+        self.zbins = np.linspace(zlo,zhi,N_zbins+1)
+        self.filters = [(lambda z: (z > zlo) & (z < zhi), self.z)]
+
+        #validation data
+        self.validation_data = {}
+        possible_observations = data_info.keys()
+        self.observation = observation
+        #check for valid combinations
+        if len(observation)==0:
+            print ('Warning: no data file supplied, no observation requested; only catalog data will be shown')
+        else:
+            #check that observation is known
+            if not(observation in possible_observations):
+                raise ValueError('Observation {} not available'.format(observation))
+            else:
+                #fetch validation data
+                self.validation_data = self.get_validation_data(band, observation)
+                
+        #plotting variables
+        self.normed = normed
+        self.ncolumns = int(ncolumns)
+
+        #setup subplot configuration and get magnitude cuts for each plot
+        self.mag_lo,self.mag_hi = self.init_plots(m_lo,m_hi)
+        
+        #setup summary plot
+        self.summary_fig, self.summary_ax = plt.subplots(self.nrows,self.ncolumns,figsize=(figx_p,figy_p),sharex='col')
+        self.first_pass = True    #only plot validation data once
+
+        #other
+        self._other_kwargs = kwargs
+
+        return
+
+    def init_plots(self, mlo, mhi):
+        #get magnitude cuts based on validation data or default limits
+        mag_lo = self.validation_data.get('mag_lo', [float(m) for m in range(mhi, mlo+1)])
```

Comment:
do the default values actually make sense? If there is not `mag_lo` and `mag_hi` in the validation data, I don't think the code would run... 

suggestion: remove them

Reply:
Yes, because we might be plotting bands for which there is no observational data.
See above